src/ralph/virtual/management/commands/openstack_sync.py

```diff
@@ -0,0 +1,532 @@
+            ralph_names = self.ralph_flavors[flavor_id]['components']
+        except KeyError:
+            ralph_names = set()
+
+        for component in set(ralph_names) - set(names):
+            model = ComponentModel.objects.get(name=component)
+            self._delete_object(GenericComponent.objects.get(
+                asset=flavor_obj,
+                model=model))
+
+        if set(ralph_names) != set(names):
+            return 1
+        else:
+            return 0
+
+    def _add_cpu(self, flavor, flavor_obj):
+        """ add cpu """
+        name = "{} cores vCPU".format(flavor['cores'])
+        try:
+            model = ComponentModel.objects.get(name=name)
+        except ObjectDoesNotExist:
+            model = ComponentModel(
+                name=name,
+                cores=flavor['cores'],
+                family="vCPU",
+                type=ComponentType.processor
+            )
+            self._save_object(model, 'Add new vCPU component model')
+
+        try:
+            GenericComponent.objects.get(asset=flavor_obj, model=model)
+        except ObjectDoesNotExist:
+            self._save_object(
+                GenericComponent(asset=flavor_obj, model=model),
+                "Add vCPU component"
+            )
+
+        return name
+
+    def _add_ram(self, flavor, flavor_obj):
+        """ add ram"""
+        name = "{} MiB vMEM".format(flavor['ram'])
+        try:
+            model = ComponentModel.objects.get(name=name)
+        except ObjectDoesNotExist:
+            model = ComponentModel(
+                name=name,
+                size=flavor['ram'],
+                type=ComponentType.memory,
+            )
+            self._save_object(model, "Add memory component model")
+
+        try:
+            GenericComponent.objects.get(asset=flavor_obj, model=model)
+        except ObjectDoesNotExist:
+            self._save_object(
+                GenericComponent(asset=flavor_obj, model=model),
+                "Add memory component"
+            )
+
+        return name
+
+    def _add_disk(self, flavor, flavor_obj):
+        """ add ram"""
+        name = "{} GiB vHDD".format(flavor['disk'])
+        try:
+            model = ComponentModel.objects.get(name=name)
+        except ObjectDoesNotExist:
+            model = ComponentModel(
+                name=name,
+                size=flavor['disk']*1024,
+                type=ComponentType.disk,
+            )
+            self._save_object(model, 'Add new disk component model')
+
+        try:
+            GenericComponent.objects.get(asset=flavor_obj, model=model)
+        except ObjectDoesNotExist:
+            self._save_object(
+                GenericComponent(asset=flavor_obj, model=model),
+                "Add disk component"
+            )
+
+        return name
+
+    def _add_flavor(self, flavor, flavor_id):
+        """Add/modify flavor ro ralph"""
+        if flavor_id not in self.ralph_flavors:
```

Comment:
why do you have to collect all flavors before this method? what if some other process add another flavor in the meantime? I'll suggest sth like this:
```
mod = False
ralph_flavor, created = CloudFlavor.objects.get_or_create(flavor_id=flavor_id, cloudprovider=self.cloud_provider, defaults=dict(name=flavor['name']))
if ralph_flavor.name != flavor['name']:
    ralph_flavor.name = flavor['name']
    self._save_object(ralph_flavor, 'Change name')
    mod = True
mod = self._add_components(flavor, ralph_flavor) or mod
if created:
    self.summary['new_flavors'] += 1
elif mod:
    self.summary['mod_flavors'] += 1
```
(unfortunately this `get_or_create` will probably doesn't work with your reversion mechanism)